lib/logger.js

```diff
@@ -6,449 +8,449 @@
+  } else if (!this._index_meta) {
+    message.meta = stringify(message.meta)
+  }
 
-    if (!this._isLoggingBackedOff && (this._bufByteLength >= this._flushLimit)) {
-        debug('Buffer size meets (or exceeds) flush limit.  Immediately flushing');
-        this._flush((err) => {
-            if (err) {
-                debug('Received an error while flushing...' + err);
-            }
-        });
-    }
+  if (this._err) {
+    return callback(this._err)
+  }
 
-    if (this._isLoggingBackedOff) {
-        debug('Backing off.');
-        this._isLoggingBackedOff = false;
-        this._flusher = setTimeout(() => {
-            this._flush(callback);
-        }, this._retryTimeout);
-    }
+  this._bufferLog(message, callback)
+}
 
-    if (!this._flusher) {
-        debug('No scheduled flush. Scheduling for %d ms from now.', configs.FLUSH_INTERVAL);
-        this._flusher = setTimeout(() => {
-            this._flush(callback);
-        }, this._flushInterval);
-    }
-};
+Logger.prototype._bufferLog = function(message, callback) {
+  if (!callback || typeof callback !== 'function') {
+    callback = (err) => { debug(err) }
+  }
+
+  if (!message || !message.line) {
+    debug('Ignoring empty message')
+    return setImmediate(callback)
+  }
+
+  if (this._max_length && message.line.length > configs.MAX_LINE_LENGTH) {
+    message.line = message.line.substring(0, configs.MAX_LINE_LENGTH) + ' (cut off, too long...)'
+    debug('Line was longer than ' + configs.MAX_LINE_LENGTH + ' chars and was truncated.')
+  }
+
+  this._bufByteLength += sizeof(message)
+
+  if (this._bufByteLength >= this._failedBufRetentionLimit) {
+    debug(`buffer size exceeded the failed buffer retention limit. Dropping this line: ${message.line}`)
+    this._bufByteLength -= sizeof(message)
+  } else {
+    debug('Buffering message: %s', message.line)
+    this._buf.push(message)
+  }
+
+  if (!this._isLoggingBackedOff && (this._bufByteLength >= this._flushLimit)) {
+    debug('Buffer size meets (or exceeds) flush limit.  Immediately flushing')
+    this._flush(callback)
+    return
+  }
+
+  if (this._isLoggingBackedOff) {
+    debug('Backing off.')
+    this._isLoggingBackedOff = false
+    this._flusher = setTimeout(() => {
+      this._flush(callback)
+    }, this._retryTimeout)
+  }
+
+  if (!this._flusher) {
+    debug('No scheduled flush. Scheduling for %d ms from now.', configs.FLUSH_INTERVAL)
+    this._flusher = setTimeout(() => {
+      this._flush(callback)
+    }, this._flushInterval)
+  }
+}
 
 Logger.prototype.addMetaProperty = function(key, value) {
-    this._meta[key] = value;
-};
+  this._meta[key] = value
+}
 
 Logger.prototype.removeMetaProperty = function(key) {
-    if (this._meta[key]) {
-        delete this._meta[key];
-    } else {
-        throw new Error(`There is no meta property called ${key}`);
-    }
-};
+  if (this._meta[key]) {
+    delete this._meta[key]
+  } else {
+    throw new Error(`There is no meta property called ${key}`)
```

Comment:
I know this is not `patch` change but throwing an error in this case is a bit aggressive - I mean, `delete` doesn't cause an issue if there is no `key` property of `meta`, right?

Reply:
You're right, but IMO, this is  method that's called manually, so throwing makes sense.  I think "noop" operations in general are confusing.  If it doesn't throw an error, how is the caller supposed to know that their request was ignored?

Reply:
@darinspivey, let's leave it like this for now but I'm worrying about the customers complaining about this is crashing their systems just because of non-existent key - we can, at least, print warning log.

Reply:
Ok, sure.  This code is old, though. It's only in here due to linting.

Reply:
i remember there was discussion about using delete not being ideal or performant:

https://flaviocopes.com/how-to-remove-object-property-javascript/

Reply:
Correct, @dm36 , we should never use `delete` because it causes map transitions. That will be addressed the next PR: https://logdna.atlassian.net/browse/LOG-6626